lib/Dialect/AIG/Transforms/LowerVariadic.cpp

```diff
@@ -0,0 +1,112 @@
+//===- LowerVariadic.cpp ---------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass lowers variadic AndInverter operations to binary AndInverter
+// operations.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/AIG/AIGOps.h"
+#include "circt/Dialect/AIG/AIGPasses.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "aig-lower-variadic"
+
+namespace circt {
+namespace aig {
+#define GEN_PASS_DEF_LOWERVARIADIC
+#include "circt/Dialect/AIG/AIGPasses.h.inc"
+} // namespace aig
+} // namespace circt
+
+using namespace circt;
+using namespace aig;
+
+//===----------------------------------------------------------------------===//
+// Conversion patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+static Value lowerFullyAssociativeOp(AndInverterOp op, OperandRange operands,
+                                     ArrayRef<bool> inverts,
+                                     PatternRewriter &rewriter) {
+  Value lhs, rhs;
+  switch (operands.size()) {
+  case 0:
+    assert(0 && "cannot be called with empty operand range");
+    break;
+  case 1:
+    if (inverts[0])
+      return rewriter.create<AndInverterOp>(op.getLoc(), operands[0], true);
+    else
+      return operands[0];
+  case 2:
+    lhs = operands[0];
+    rhs = operands[1];
+    return rewriter.create<AndInverterOp>(op.getLoc(), lhs, rhs, inverts[0],
+                                          inverts[1]);
+  default:
+    auto firstHalf = operands.size() / 2;
+    lhs = lowerFullyAssociativeOp(op, operands.take_front(firstHalf),
```

Comment:
This seems fine for proof of concept, but is there a potential for stack overflow with large inputs if we recurse along use-def edges?

Reply:
The recursion depth is O(log N) where N is operand size so it should be fine (actually we perform the same transformation at PrepareForEmission).

Reply:
I see, fair enough.